TSOClient/tso.simantics/NetPlay/EODs/Handlers/VMEODHoldEmCasinoPlugin.cs

```diff
@@ -0,0 +1,2187 @@
+                        return;
+                    }
+
+                    // add to lobby
+                    if (Lobby.Join(client, playerIndex))
+                    {
+                        var slot = Lobby.GetSlotData(client);
+                        if (slot != null)
+                        {
+                            // get the amount of money the player has by sending a testOnly transaction for $1 from the table
+                            var VM = client.vm;
+
+                            VM.GlobalLink.PerformTransaction(VM, true, Server.Object.PersistID, client.Avatar.PersistID, 1,
+                                (bool success, int transferAmount, uint uid1, uint budget1, uint uid2, uint budget2) =>
+                                {
+                                    if (success)
+                                    {
+                                        // set the slot data
+                                        slot.Client = client;
+                                        slot.PlayerIndex = playerIndex;
+                                        slot.SimoleonBalance = (int)budget2;
+                                        slot.ResetHand();
+                                        slot.OnPlayerAnteBetChange += BroadcastSingleAnteBet;
+                                        slot.OnPlayerSideBetChange += BroadcastSingleSideBet;
+
+                                        // set the table data
+                                        TableBalance = (int)budget1;
+
+                                        if (IsTableWithinLimits())
+                                        {
+                                            string[] data = new string[] { "" + playerIndex, MinAnteBet + "", MaxAnteBet + "", MaxSideBet + "",
+                                                "" + Dealer.Avatar.ObjectID };
+
+                                            // event to show the UI to the player
+                                            client.Send("holdemcasino_player_show", VMEODGameCompDrawACardData.SerializeStrings(data));
+
+                                            if (Dealer != null)
+                                            {
+                                                if (GameState.Equals(VMEODHoldEmCasinoStates.Waiting_For_Player) || GameState.Equals(VMEODHoldEmCasinoStates.Closed))
+                                                    EnqueueGotoState(VMEODHoldEmCasinoStates.Betting_Round);
+                                                else // player is joining mid-game or mid-betting round
+                                                {
+                                                    if (GameState.Equals(VMEODHoldEmCasinoStates.Betting_Round))
+                                                    {
+                                                        // sync all ante and side bets
+                                                        List<string> acceptedBets = GetAllAcceptedBets();
```

Comment:
This is not thread safe as it iterates Lobby.Players, which can change on the main thread.
As a rule, anything that isn't a single primitive number could partially change when running on another thread. Primitive number types change atomically, so your problem there is thinking if multiple could change that depend on each other's value remaining consistent.

This one is quite tricky to solve. I'm OK with you ignoring this, but it would be nicer if you somehow made these actions happen on the main thread. (usually I use as lock guarded List<Action> to run arbitrary functions on the main thread)

Reply:
In fact, this entire PerformTransaction handler could use some running on the main thread, it's very hard to reason about this much access to shared state. Might be worth considering a standard way to do this between all EOD servers using helper functions.

Reply:
Definitely agree, fixes all around.